File: ABC/001-050/ABC-016/ABC016-C.py
N, M = map(int, input().split())
graph = {i: set() for i in range(1, N + 1)}
num = []
for i in range(M):
    a, b = map(int, input().split())
    num.append((a, b))

for a, b in num:
    graph[a].add(b)
    graph[b].add(a)

# IDがNの人の友達の友達の人数を返す関数
def friends_of_friends(N):
    friends_list = set()
    friends_of_friends_list = set()
    
    # 直接の友達をfriends_listに追加
    for i in graph[N]:
        friends_list.add(i)
    
    # 友達の友達をfriends_of_friends_listに追加
    for friend in friends_list:
        for friend_of_friend in graph[friend]:
            if friend_of_friend != N and friend_of_friend not in friends_list:
                friends_of_friends_list.add(friend_of_friend)
    
    return len(friends_of_friends_list)

for i in range(1, N + 1):
    print(friends_of_friends(i))

File: ABC/001-050/ABC-016/ABC016-C.rs
use std::collections::{HashSet, HashMap};
use std::io;

fn main() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Failed to read input");
    let mut iter = input.split_whitespace();
    let n: usize = iter.next().unwrap().parse().expect("Invalid input");
    let m: usize = iter.next().unwrap().parse().expect("Invalid input");

    let mut graph: HashMap<usize, HashSet<usize>> = HashMap::new();
    for i in 1..=n {
        graph.insert(i, HashSet::new());
    }

    let mut num: Vec<(usize, usize)> = Vec::new();
    for _ in 0..m {
        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("Failed to read input");
        let mut iter = input.split_whitespace();
        let a: usize = iter.next().unwrap().parse().expect("Invalid input");
        let b: usize = iter.next().unwrap().parse().expect("Invalid input");
        num.push((a, b));
    }

    for (a, b) in num.iter() {
        graph.get_mut(a).unwrap().insert(*b);
        graph.get_mut(b).unwrap().insert(*a);
    }

    for i in 1..=n {
        println!("{}", friends_of_friends(i, &graph));
    }
}

fn friends_of_friends(id: usize, graph: &HashMap<usize, HashSet<usize>>) -> usize {
    let mut friends_list: HashSet<usize> = HashSet::new();
    let mut friends_of_friends_list: HashSet<usize> = HashSet::new();

    if let Some(friends) = graph.get(&id) {
        for &friend in friends {
            friends_list.insert(friend);
        }

        for &friend in friends_list.iter() {
            if let Some(friends_of_friend) = graph.get(&friend) {
                for &friend_of_friend in friends_of_friend {
                    if friend_of_friend != id && !friends_list.contains(&friend_of_friend) {
                        friends_of_friends_list.insert(friend_of_friend);
                    }
                }
            }
        }
    }

    friends_of_friends_list.len()
}